pkg/operator/finalizer.go

```diff
@@ -0,0 +1,167 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+
+	regopapi "github.com/openshift/cluster-image-registry-operator/pkg/apis/imageregistry/v1alpha1"
+	"github.com/openshift/cluster-image-registry-operator/pkg/generate"
+	"github.com/openshift/cluster-image-registry-operator/pkg/parameters"
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+)
+
+func (h *Handler) RemoveResources(o *regopapi.ImageRegistry) error {
+	modified := false
+
+	templetes, err := h.GenerateTemplates(o, &h.params)
+	if err != nil {
+		return fmt.Errorf("unable to genetate templates: %s", err)
+	}
+
+	for _, tmpl := range templetes {
+		err = generate.RemoveByTemplate(tmpl, &modified)
+		if err != nil {
+			return fmt.Errorf("unable to remove objects: %s", err)
+		}
+		logrus.Infof("resource %s removed", tmpl.Name())
+	}
+
+	configState, err := generate.GetConfigState(h.params.Deployment.Namespace)
+	if err != nil {
+		return fmt.Errorf("unable to get previous config state: %s", err)
+	}
+
+	err = generate.RemoveConfigState(configState)
+	if err != nil {
+		return fmt.Errorf("unable to remove previous config state: %s", err)
+	}
+
+	return nil
+}
+
+func (h *Handler) finalizeResources(o *regopapi.ImageRegistry) error {
+	defer utilruntime.HandleCrash()
+
+	if o.ObjectMeta.DeletionTimestamp == nil {
+		return nil
+	}
+
+	finalizers := []string{}
+	for _, v := range o.ObjectMeta.Finalizers {
+		if v != parameters.ImageRegistryOperatorResourceFinalizer {
+			finalizers = append(finalizers, v)
+		}
+	}
+
+	if len(finalizers) == len(o.ObjectMeta.Finalizers) {
+		return nil
+	}
+
+	logrus.Infof("finalizing %s %s/%s", o.GetObjectKind().GroupVersionKind().Kind, o.Namespace, o.Name)
+
+	err := h.RemoveResources(o)
+	if err != nil {
+		return err
+	}
+
+	cr := o
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if cr == nil {
+			cr = &regopapi.ImageRegistry{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: o.TypeMeta.APIVersion,
+					Kind:       o.TypeMeta.Kind,
+				},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      o.ObjectMeta.Name,
+					Namespace: o.ObjectMeta.Namespace,
+				},
+			}
+
+			err := sdk.Get(cr)
+			if err != nil {
+				return fmt.Errorf("failed to get custom resource %s/%s: %s", o.ObjectMeta.Namespace, o.ObjectMeta.Name, err)
+			}
+		}
+
+		cr.ObjectMeta.Finalizers = finalizers
+
+		err := sdk.Update(cr)
+		if err != nil {
+			cr = nil
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("unable to update finalizers in custom resource %s/%s: %s", o.ObjectMeta.Namespace, o.ObjectMeta.Name, err)
+	}
+
+	// These errors may indicate a transient error that we can retry in tests.
+	errorFuncs := []func(error) bool{
+		kerrors.IsInternalError,
+		kerrors.IsTimeout,
+		kerrors.IsServerTimeout,
+		kerrors.IsTooManyRequests,
+		utilnet.IsProbableEOF,
+		utilnet.IsConnectionReset,
+	}
+
+	retryTime := 3 * time.Second
+
+	err = wait.PollInfinite(retryTime, func() (stop bool, err error) {
```

Comment:
why do we need to wait for the deletion to happen?

Reply:
We do not allow two resources in the same space.

Reply:
hm.  So you are waiting here until it goes away before you (re)create the default samplesresource?

Reply:
err, image registry resource.

Reply:
> So you are waiting here until it goes away before you (re)create the default samplesresource?

I'm waiting for the removal of this particular resource so that you can create another resource and not finalize this one again (we can miss a delete event by definition).

Reply:
makes sense, thanks